hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/upgrade/UpgradeManager.java

```diff
@@ -0,0 +1,204 @@
+
+  public void run(OzoneConfiguration configuration) throws IOException {
+    final DatanodeDetails detail =
+        UpgradeUtils.getDatanodeDetails(configuration);
+
+    final MutableVolumeSet dataVolumeSet = UpgradeUtils
+        .getHddsVolumes(configuration, StorageVolume.VolumeType.DATA_VOLUME,
+            detail.getUuidString());
+
+    for (StorageVolume storageVolume : dataVolumeSet.getVolumesList()) {
+      final HddsVolume volume = (HddsVolume) storageVolume;
+      volume.loadDbStore(false);
+    }
+    initVolumeStoreMap(dataVolumeSet, configuration);
+    upgradeAll(dataVolumeSet, configuration);
+  }
+
+  public void initVolumeStoreMap(MutableVolumeSet dataVolumeSet,
+                                 OzoneConfiguration configuration)
+      throws IOException {
+    for (StorageVolume storageVolume : dataVolumeSet.getVolumesList()) {
+      final HddsVolume volume = (HddsVolume) storageVolume;
+      final File containerDBPath = getContainerDBPath(volume);
+      final DatanodeStoreSchemaThreeImpl datanodeStoreSchemaThree =
+          new DatanodeStoreSchemaThreeImpl(configuration,
+              containerDBPath.getAbsolutePath(), false);
+      volumeStoreMap.put(volume.getStorageDir().getAbsolutePath(),
+          datanodeStoreSchemaThree);
+    }
+  }
+
+  public List<Result> upgradeAll(MutableVolumeSet volumeSet,
+                                 OzoneConfiguration configuration) {
+    List<Result> results = new ArrayList<>();
+    final List<StorageVolume> volumesList = volumeSet.getVolumesList();
+    List<CompletableFuture<Result>> volumeFutures =
+        new ArrayList<>(volumesList.size());
+    long startTime = System.currentTimeMillis();
+
+    LOG.info("Start upgrade {} volumes container LayoutVersion",
+        volumesList.size());
+
+    for (StorageVolume volume : volumesList) {
+      final HddsVolume hddsVolume = (HddsVolume) volume;
+      final UpgradeTask upgradeVolume =
+          new UpgradeTask(configuration, hddsVolume,
+              volumeStoreMap.get(hddsVolume.getStorageDir()
+                  .getAbsolutePath()));
+
+      final CompletableFuture<Result> future =
+          upgradeVolume.getUpgradeFutureByVolume();
+      volumeFutures.add(future);
+    }
+
+    for (CompletableFuture<Result> volumeFuture : volumeFutures) {
+      final Result result = volumeFuture.join();
+      results.add(result);
+      LOG.info("Finish upgrade containers on volume {}, result {}",
+          result.getHddsVolumeRootDir(), result);
+    }
+
+    LOG.info("Upgrade all volume container LayoutVersion costs {}s",
+        (System.currentTimeMillis() - startTime) / 1000);
+    return results;
+  }
+
+  @VisibleForTesting
+  public DatanodeStoreSchemaThreeImpl getDBStore(HddsVolume volume) {
+    return volumeStoreMap.get(volume.getStorageDir().getAbsolutePath());
+  }
+
+  /**
+   * This class response upgrade v2 to v3 container result.
+   */
+  public static class Result {
```

Comment:
Is it able to merge the two Result class using one？

Reply:
> Is it able to merge the two Result class using one？

One of them is the result of the container upgrade and the other is a wrapper of all the container upgrade results in the volume, let me rename one of them